yi-core/src/Yi/Option.hs

```diff
@@ -0,0 +1,125 @@
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE Rank2Types #-}
+{-# OPTIONS_HADDOCK show-extensions #-}
+
+-- |
+-- Module      :  Yi.Option
+-- License     :  GPL-2
+-- Maintainer  :  yi-devel@googlegroups.com
+-- Stability   :  experimental
+-- Portability :  portable
+--
+-- Command-line options
+
+module Yi.Option
+    (
+    -- * Types
+      YiOption
+    , YiOptionDescr
+    , YiOptions
+    , OptionError(..)
+
+    -- * Core
+    , yiCustomOptions
+    , consYiOption
+    , consYiOptions
+
+    -- * Argument-less options
+    , yiBoolOption
+    , yiFlagOption
+    , yiActionFlagOption
+
+    -- * Argument-taking options
+    , yiStringOption
+    , yiStringOption'
+    , yiActionOption
+    , yiActionOption'
+    )
+where
+
+import           Data.Default          (Default)
+import qualified Data.Text             as T (Text)
+import           Data.Typeable         (Typeable)
+import           Data.String           (IsString, fromString)
+import           Lens.Micro.Platform   (Lens', makeLenses, over, set)
+import           System.Exit           (ExitCode)
+import           System.Console.GetOpt (OptDescr, ArgDescr(..))
+import           Yi.Config.Lens        (configVariable, startActionsA)
+import           Yi.Types              (Action, Config, YiConfigVariable)
+
+data OptionError = OptionError T.Text ExitCode
+
+-- | An option is a function that attempts to change the configuration of the
+-- editor at runtime.
+type YiOption = Config -> Either OptionError Config
+
+type YiOptionDescr = OptDescr YiOption
+
+-- | Custom options that should be accepted. Provided in user configuration.
+--
+-- The general flow is that the user adds options to his configuration. Options
+-- are essentially functions describing how to modify the configuration at runtime.
+-- When an option is called, it gets the current config and may modify it (to encode
+-- its value)
+newtype YiOptions = YiOptions { _yiOptions :: [YiOptionDescr] }
+    deriving (Default, Typeable)
```

Comment:
[deriving `Typeable` is not necessary](https://downloads.haskell.org/~ghc/7.10.3/docs/html/users_guide/deriving.html#deriving-typeable)